### ISSUE TITLE
Add -deepFolders to PathName

### ISSUE THREAD

**Issue by madskjeldgaard** [supercollider/supercollider#5758]
## Motivation
PathName has a neat method called deepFiles which finds all the files in a file tree. A folder equivalent would be really convenient and so I would like to propose we make a `deepFolders` method.

## Description of Proposed Feature
This simple proposotion is a literal mirror of the deepFiles method. It seems to work nicely:

```
+PathName{
    deepFolders {
        ^this.entries.collect({ | item |
            if(item.isFolder, {
                item
            },{
                item.deepFolders
            })
        }).flat
    }
}
```
## Plan for Implementation
Put the above method inside of PathName.sc

**Comment by jamshark70** [comment]
I think it needs to be written this way instead:

```
+ PathName {
	deepFolders {
		var out;
		this.entries.do { | item |
			if(item.isFolder) {
				out = out.add(item);
				out = out.addAll(item.deepFolders);
			}
		};
		^out
	}
}
```

The logic in the original post is "if it is a folder, add it to the results; if it is not a folder, scan for sub-entries" **but** if the item is not a folder, then it doesn't have sub-entries.

I tested with a nested folder structure in my filesystem and, as I suspected, I got the wrong result of only the top-level folders.

deepFolders needs to put the recursive branch into the `isFolder` branch.

This makes it inconvenient to use `collect` because this method always adds to the result, but we want to ignore non-folders. You could `.flat.select(_.notNil)` but then the filtering will be performed at every level of recursion, which is wasteful. I think it's more efficient to add only when necessary. (`addAll` seems to flatten inherently! A nice bonus.)

**Comment by madskjeldgaard** [comment]
> + PathName {
> 	deepFolders {
> 		var out;
> 		this.entries.do { | item |
> 			if(item.isFolder) {
> 				out = out.add(item);
> 				out = out.addAll(item.deepFolders);
> 			}
> 		};
> 		^out
> 	}
> }

You're totally right. This works much better. Just tested it here - it's great! Do you want to create a PR for it or should I? I think this is a super useful method to have
